include/seastar/core/abort_source.hh

```diff
@@ -100,15 +100,16 @@ public:
 
 private:
     using subscription_list_type = bi::list<subscription, bi::constant_time_size<false>>;
-    std::optional<subscription_list_type> _subscriptions = subscription_list_type();
+    subscription_list_type _subscriptions;
     std::exception_ptr _ex;
 
     void do_request_abort(std::optional<std::exception_ptr> ex) noexcept {
-        assert(_subscriptions);
+        if (_ex) {
+            return;
+        }
         _ex = ex.value_or(get_default_exception());
-        auto subs = std::exchange(_subscriptions, std::nullopt);
-        while (!subs->empty()) {
-            subscription& s = subs->front();
+        while (!_subscriptions.empty()) {
+            subscription& s = _subscriptions.front();
```

Comment:
`on_abort` might delete `this` abort_source (like in the semaphore entry case).
So we need something like the following to work on a local copy of the subscriptions list:
```
        auto subs = std::move(_subscriptions);
        while (!subs.empty()) {
```
instead of `auto subs = std::exchange(_subscriptions, std::nullopt)` that was done before.